### ISSUE TITLE
Evaluate NDepend (OSOE-135)

### ISSUE THREAD

**Issue by Piedone** [Lombiq/Open-Source-Orchard-Core-Extensions#133]
[NDepend](https://www.ndepend.com/) is a .NET code quality tool. We've received 4 licenses due to the generosity of NDepend's founder, Patrick Smacchia. Let's try NDepend on OSOCE:

- Show what kind of issues it finds and evaluate whether there are features in NDepend we can benefit from using in the long-term.
- Create issues for the findings in each repo.
- Demonstrate running it in CI with GitHub Actions if it's possible.
- Summarize our experiences here in form of a comment in about 3-4 paragraphs.

**Comment by Piedone** [comment]
Patrick, if you can reply here: Do I see correctly that there's no NDepend GitHub Actions integration readily available? Under [this page](https://www.ndepend.com/docs/ndepend-github-action), the video is about Azure DevOps and the link pointing to the GitHub Marketplace is a dead one. I didn't find anything else either.

**Comment by Piedone** [comment]
Here are my findings:
- NDepend produces a very nice report.
- The code quality rules are some similar rules to what we already have in our [.NET Analyzers](https://github.com/Lombiq/.NET-Analyzers) project, e.g. “Avoid types too big”, “Avoid methods with too many parameters”. We already have tuned these to what we want to allow.
- Suspected bugs I've found:
  - “Types with disposable instance fields must be disposable” doesn’t take `IAsyncDisposable` into account. See e.g. [SmtpService](https://github.com/Lombiq/UI-Testing-Toolbox/blob/dev/Lombiq.Tests.UI/Services/SmtpService.cs).
  - Reported “Avoid interfaces too big“ for [`IOrchardServices<T>`](https://github.com/Lombiq/Helpful-Libraries/blob/dev/Lombiq.HelpfulLibraries.OrchardCore/DependencyInjection/IOrchardServices.cs). Couldn’t open it from the Queries and Rules Edit windows due to this error:
![image-20220807-203136](https://user-images.githubusercontent.com/1976647/183450534-e6367fad-4c85-4d88-995d-4bded7e5e8b8.png)
  - Treemap Metric View uses the generated C# Razor code for Razor templates (cshtml files). While technically this is correct, I don’t think it’s too useful.
- The VS extension works really well.
- Probably the biggest advantage over our existing analyzers when it comes to code quality checks is the nice report, which is available in HTML and from within VS in an interactive manner too. That it tracks historical trends (i.e. we can see if our code is getting better or worse) is a plus too, though how we use analyzers is that all violations cause a build to fail, so there’s no code merged that still needs improvement like this (other metrics would be interesting to track though). The downsides are (apart from us already having analyzers integrated and fine-tuned) the lack of live feedback in the IDE, the lack of GitHub Actions integration, and that it’s not open-source but priced per user.

I don’t think we need to be worried too much about the issues uncovered by NDepend. A lot of them are due to deliberate design decisions, some that other analyzers caught too but we suppressed them.

Apart from code quality rules, NDepend also provides a Dependency Graph (this is what we want from a tool called “NDepend”!), Dependency Matric, and Treemap Metric View. These are useful, easy-to-consume visualizations of dependencies and the amount of code in components of the solution, respectively (and the primary reason why I wanted to check out NDepend!). They are only possible to consume in the VS extension, however, the HTML report only provides small, illegible, and static images of them (at least for OSOCE).
